### ISSUE TITLE
Snowflake: Inconsistent column name case

### ISSUE THREAD

**Issue by rumbin** [apache/superset#18085]
Superset handles the case of non-case-sensitive column names inconsistently for Snowflake connections:

* Native tables/views, i.e., physical datasets, have _lowercase_ column names, as it is internally handled by SQLAlchemy.
* SQLLab's virtual datasets have _UPPERCASE_ column names.

This is troublesome for all dashboards where filters are present which act on charts with related, mixed physical/virtual datasets.  
Superset's filter scoping is case sensitive. Thus, filters on a certain column name will _either_ be applied to the related physical _or_ virtual datasets.

#### How to reproduce the bug

1. Register any Snowflake table with non-case-sensitive column names (i.e., UPPERCASE) as a physical dataset in Superset.
2. Notice how its columns are recognized and stored in lowercase.
3. Edit the dataset again and convert it to a _virtual_ dataset. Use `select * from …` as statement text. Save the changes.
4. Edit the dataset again and synchronize the column names.
5. Notice how the column names now changed to UPPERCASE. 

### Expected results

The case should not depend on whether the dataset is physical or virtual.  
All non-case-sensitive (i.e., UPPERCASE in Snowflake) should be converted to lowercase consistently, also for virtual datasets created in SQLLab.

Lowercase is the internal representation of SQLALchemy.

### Actual results
Virtual dataset's column names are treated as UPPERCASE.

#### Screenshots
The easies way to experience this effect is to simply explore a Snowflake table in SQLLab.
As you can see in the following screenshot, the column names that are extracted from the schema and displayed in the schema browser on the left are represented as _lowercase_, while the query results of the preview table have _UPPERCASE_ column names:

![image](https://user-images.githubusercontent.com/1220356/150071559-d51466fe-e181-4148-8628-a6b4d4b17f43.png)

### Rejected workarounds
An option would of course be to use double-quoted lowercase column aliases in the `SELECT` statement of the virtual dataset.
This would make these column names case-sensitive and thus be treated as lowercase.
However, with a broad userbase working with Snowflake/Superset this is very unhandy.  
Furthermore, this apporach would not allow any `select *` statements here.

I think it simply is a bug and it needs to be fixed so the way the column names are treated is always consistent.

### Environment

Tested versions:

* 1.0.1
* 1.4 rc4

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### Additional context

This SQLALchemy issue is closely related: https://github.com/snowflakedb/snowflake-sqlalchemy/issues/157#issuecomment-807922786

However, my feeling is that we should probably try to stick with SQLALchemy's way of treating the column names as lowercase instead of uppercase. However, then we need to do so consistently and correct this for the virtual datasets.

**tai** pointed me in Slack to this piece of code: https://github.com/snowflakedb/snowflake-sqlalchemy/blob/9118cf8f18a0039f9cb5d3892ff2b1e5c82a05e0/snowdialect.py#L217

**Comment by rumbin** [comment]
Would it be feasible to have the case conversion configurable for each database connection?

I am thinking about ways to fix this without breaking existing datasets/charts/dashboards that would suddenly have their colunm name case changed upon resync of the column names...

**Comment by nytai** [comment]
This issue is the same as the one you linked https://github.com/snowflakedb/snowflake-sqlalchemy/issues/157#issuecomment-807922786 


Superset is using that library to connect to snowflake so it needs to be fixed there, there’s nothing that can be done in superset to fix this.

**Comment by rumbin** [comment]
@nytai I do not agree that this is the same issue.
In the upstream issue it is proposed to treat the columns as uppercase while the proposed solution here is lowercase.

I think that there needs to be a design discussion in this issue before deciding how the column case is to be handled.
This _may_ then lead to the conclusion that the upstream bug is sufficient.

Would you please reopen the issue?

**Comment by rumbin** [comment]
@nytai The linked upstream issue is referring to another issue that had been closed as the issue can be handled by the client's code: https://github.com/snowflakedb/snowflake-sqlalchemy/issues/157#issuecomment-913699495 

@villebro was involved in closing the related upstream issue: https://github.com/snowflakedb/snowflake-sqlalchemy/issues/73

So it seems to me as a bit of a ping pong to where this issue is to be handled.

I'd be glad if Ville could comment on this. Maybe we could also have a Slack discussion of video call if needed.

**Comment by rumbin** [comment]
From the linked upstream issues I understand that it is unlikely that this porblem will be fixed upstream.

I understood from these issues that the correct case can be extracted from the cursor, but it seems that we are not currently doing so in Superset.
Furthermorte, I am not quite sure what the desired behavior in Superset would actually be.

So, in my eyes there first needs to be a design decision.

**Comment by nytai** [comment]
ok, revisiting this it seems unlikely that that snowflake-sqlalchemy will stop denormalizing the column names in the payload given how long it's been open and that the cursor description seems to be an acceptable workaround. 

I now agree that this could be addressed in superset given the challenges it presents on dashboards with charts that are backed by virtual & physical datasets.

**Comment by villebro** [comment]
Ok, so this all stems from the design decision Oracle-like databases have made, which I generally call "lowercase means uppercase means caseless", which IMO is a terrible assumption (I _really_ wish Snowflake hadn't gone with this design). I may be off on some details, but basically it boils down to this: When you write

```sql
select column
from table
```

and the column name is caseless (=stored in metadata as UPPERCASE), the resulting column name in the query result is `COLUMN`. You get the same result by writing (correct me if I misremember these)

```sql
select Column
from table
```

or

```sql
select COLUMN
from table
```

or even

```sql
select "COLUMN"
from table
```

but an error when writing

```sql
select "column"
from table
```

because the column name isn't defined case-sensitively as "column", but rather stored in the metadata as "COLUMN" (=caseless). However, when you check the column name using `inspector.get_columns`, the column shows up as `column`, i.e. lowercase, which kind of isn't really true using Oracle logic, but ends up working in Superset. The Snowflake dialect does these to/from conversions here: https://github.com/snowflakedb/snowflake-sqlalchemy/blob/9118cf8f18a0039f9cb5d3892ff2b1e5c82a05e0/snowdialect.py#L217-L234 , and these functions kind of make sense when you know the assumptions behind case handling, but is super confusing when you're trying to write code that handles the cases correctly.

Now, we actually have a method in `BaseEngineSpec` called `get_column_spec` which takes the source of the column definition as one argument: https://github.com/apache/superset/blob/035638c95869d0d4b7c9d44f13e88e3535ecdf4c/superset/db_engine_specs/base.py#L1313-L1320
What we could do here is add a field `normalized_name` to the `ColumnSpec` type and return the normalized column name on affected dbs (Oracle, Snowflake et) if the column metadata originates from a query (=virtual table), but leave it unchanged for physical tables. This would just need to be implemented in the SQLAlchemy model and shouldn't be lots of work, but it would be really important to do super thorough testing to make sure it handles all cases correctly (caseless, UPPERCASE, "lowercase" and "Mixed Case").

This probably sounds really confusing, so maybe it makes sense to have a kickoff call to get this work started to make sure we fully understand the problem and have a clear proposal for solving this.

**Comment by rumbin** [comment]
@villebro in my eyes, your elaboration on the case handling is correct (including your opinion on the poor default choice of UPPERCASE...).

About the rest of your explanation you lost me somewhere, though.
What is — in your eyes — the intended behavior of Superset regarding the _caseless_ columns. Should they become lowercase (like the physical datasets already are) or UPPERCASE (like the virtual datasets already are)?

**Comment by villebro** [comment]
_In theory_, i think the optimal solution would be to have the column cases equal for both physical and virtual datasets. I _think_ this problem would go away if we called the normalization function on the column names when we fetch column names from the query metadata when we create a new virtual dataset. However, I believe this would break backward compatibility with previously created virtual tables, as refreshing the column metadata for pre-existing virtual tables would change the case from upper case to lower case, which would probably break all existing charts. We could, of course, do a database migration which would update all existing chart metadata, but that would be very tricky, as column names appear in lots of different controls (we also don't know what custom controls containing column names some orgs might have if they've created custom viz plugins).

Therefore, the approach I'm proposing is to rather be aware of the dataset type, and then be able to work around this problem at query creation time, i.e. do the normalization there. As the dataset metadata would be unaffected (caseless virtual dataset column names would still be UPPER CASE), this change would be backwards compatible, and therefore wouldn't require any database migrations.

However, if column cases are different for physical vs virtual tables, there is the risk that native filters will work properly on datasets of different types. So this may require some further planning.

**Comment by nytai** [comment]
I was originally thinking that what would make the most sense is for superset to use the same case that it's defined in the db (ie, most likely UPPERCASE for snowflake, but not _always_), however since the sqlalchemy dialect is doing this "normalization" there really is no way to know for sure what the case is in the db. We could update the `get_column_spec` to instead use a describe table stmt (or whatever the dialect is using to get table names) and get the column names from the cursor (not sure if this would actually work). 

However, this gets very complicated when we take into consideration all the existing datasets out there, as users who upgrade to a version with this fix would start seeing issues when they run the sync columns action and the case changes (which would break all their existing dashboard filters, etc). 

I too really wish snowflake had not gone with this UPPERCASE by default design, and I especially wish that the sqlalchemy dialect hadn't gone with this "normalize" column names approach.

**Comment by nytai** [comment]
so this is the function in sqlalchemy-snowflake that returns the table names: https://github.com/snowflakedb/snowflake-sqlalchemy/blob/9118cf8f18a0039f9cb5d3892ff2b1e5c82a05e0/snowdialect.py#L418-L487

**Comment by villebro** [comment]
After giving this some more thought, I agree that the best solution would probably be to store the object names in their native case (UPPERCASE), as that would always work with quotes. If we do that, then all physical table column definitions would change from lowercase to UPPERCASE (unless MixedCase). I'm not sure we can do a database migration to fix old datasets reliably, so to deal with backwards compatibility issues, we could potentially add a config flag to the dataset metadata, something like `legacy_mode`, which we'd set to `True` for existing Snowflake datasets, but that would not be set on new datasets. When set to `True`, column names would be set with the current normalization logic, but when disabled would return the UPPERCASE name. Then existing charts wouldn't break, and users would have the option of migrating those by hand if they so wish.

**Comment by rumbin** [comment]
@villebro this is what I had in mind, too.
The backward compatibility definitely needs some good care and I think that such a flag/switch would work nicely, although it introduces some ugly legacy.

**Comment by agusfigueroa-htg** [comment]
Hi all, 

First of all thanks for the discussion. It's very enriching, and the proposed solution would reach case consistency and solve the issue as mentioned. 

After sharing this with our team another option that popped up, **in case this only affects filters interoperability**,  is to add the option to make the filters case insensitive. 

In that case, instead of looking for columns named the same (`dataset1.colname=dataset2.colname`), we can apply the filters "case insensitive" like (`lower(dataset1.colname)=lower(dataset2.colname)`), creating a flag for it in case users want to apply it or not.

This would lay fully on the Superset side and would be transparent across different databases and what they decide to do in the future (return uppercase, lowercase, mix case or so).

**Comment by rumbin** [comment]
@agusfigueroa-htg having the column matching if dashboard filters case-insensitive is a quick win in my eyes.
However, some edge cases remain:

1. When filters are handled in Jinja, users will need to implement their own case-insensitivity.
2. When users convert a physical to a virtual dataset, or vice versa, the column case will flip, causing possible bugs. Such a conversation may be needed, e.g., when the source model now includes soft deletes which then need to be filtered out, or, the other way round, when it is decided to convert a virtual dataset into a table or view.
3. Users might be confused by the case inconsistency and they might not be aware of the case-insensitivity of the filters.

**Comment by villebro** [comment]
@agusfigueroa-htg that's a really good idea, and could be useful even for other databases. However, I do feel this is an issue we probably need to resolve, and the points raised by @rumbin are IMO all very valid. It's also good to keep in mind that this not only affects Snowflake, but even other Oracle-like dialects.

**Comment by agusfigueroa-htg** [comment]
Good to see that we are all in the same boat :) and the points presented by you two are pretty much valid - the filter one is more of a quick escape rather than a real fix. 

@villebro if I got you correctly, it makes sense to work on how the object names are stored in Superset. I don't know where that is defined in the code (I am new to the repo!), but if you have some references at a hand I am happy to help here.

**Comment by rumbin** [comment]
Stumbled on a related, merged PR: #5827

**Comment by agusfigueroa-htg** [comment]
Hi @villebro ! I assume we want to solve this on the connector side, is that correct? If I get a rough idea of where to start looking at, I hope I can contribute on taking the first steps :)

**Comment by rumbin** [comment]
This issue should not be closed by the stale bot, in my eyes.

**Comment by agusfigueroa-htg** [comment]
I support rumbin here, this issue is still around!

**Comment by villebro** [comment]
Hi, I agree this is something we should address. I actually think the best solution would be to make Superset case insensitive, as I believe very few people will need to have `column`, `CoLuMn` and `COLUMN` in the same dataset. However, this is a pretty big breaking change, so it would likely need to first be floated as a SIP, and then be implemented in 3.0 if others agree with the change. @rumbin would you be up for the task of preparing a SIP for this? This way we could have the discussion there and iron out any disagreements before starting actual work to address this.

**Comment by agusfigueroa-htg** [comment]
@rumbin can count on me if I can support anyhow!

**Comment by villebro** [comment]
@rumbin to echo @agusfigueroa-htg 's offer for support - I'm also happy to help push this effort forward, I just don't have the time/resources to drive it.

**Comment by rumbin** [comment]
@villebro, @agusfigueroa-htg please have a look at #19773

**Comment by rumbin** [comment]
@villebro did you mean to write Superset 3.0? I was hoping that we could get it into 2.0, but maybe I am too naïve here...

**Comment by villebro** [comment]
Yes, unfortunately this won't make it into 2.0, as the list of breaking changes has already been decided. However, if we do reach consensus on how to fix this, we can put it behind a feature flag during 2.0, and then potentially introduce it as the new default behavior in 3.0.

**Comment by rusackas** [comment]
Hi all! Just checking in on this thread, since it's gone silent for a year. 

Is anyone still hoping to tackle this? Or did it somehow cease to be a problem? Now's a great chance to add a feature flag for 3.0, as that would be a non-breaking change in the release being cut on or around June 15th.

**Comment by rumbin** [comment]
I am not sure if I will have time and need to work on this on the near future, as I am currently not using Snowflake, but that might change soon.
Thus, it may be beneficial to at least define the feature flag already.
I cannot tell, however, if the issue disappeared miraculously, but I doubt it, judging from the depth/origin of the problem.

**Comment by agusfigueroa-htg** [comment]
I can sadly confirm that the issue remains :(

**Comment by villebro** [comment]
We've been throwing around ideas about this with @rusackas for the last few weeks, and my proposal is to change the behavior going forward as follows:
- virtual dataset logic remains unchanged
- physical tables call `denormalize_name` on the column name. See example below:

```python
>>> from snowflake.sqlalchemy.snowdialect import SnowflakeDialect
>>> dialect = SnowflakeDialect()
>>> dialect.denormalize_name('foo')
'FOO'
>>> dialect.denormalize_name('Foo')
'Foo'
>>> dialect.denormalize_name('FOO')
'FOO'
```

This means that going forward, caseless physical column names will be stored in UPPERCASE, rather than lowercase. We'll also do something similar for Oracle (I assume the oracle dialect). This will ensure that physical and virtual datasets have the same case. Furthermore, I propose not doing any migrations to existing datasets, ensuring we don't break any existing dashboards or charts. This means, that anyone who wants to migrate old charts/dashboards to the new behavior will need to edit the dataset and click "Sync columns from source":

<img width="766" alt="image" src="https://github.com/apache/superset/assets/33317356/b47ed3b9-01f2-4615-a19e-0e85799f8ae9">

After this the lowercase column names will become UPPERCASE:

<img width="766" alt="image" src="https://github.com/apache/superset/assets/33317356/8761d5e0-4d74-40ce-b6d0-1bca04d2436e">

Thoughts @agusfigueroa-htg @rumbin @nytai ?

**Comment by rumbin** [comment]
@villebro thanks a lot for giving this issue some love :-).

The suggested approach seems like a simple yet effective solution to me.

**Comment by agusfigueroa-htg** [comment]
I second @rumbin very much here, this is a simple and effective solution. Thinking about our Superset instance, this will definitely address the problems we have with dashboard filtering.

Thank you a million @villebro @rusackas, very much appreciate this.

**Comment by crysensible** [comment]
Hi all, this change has broken all of our internal charts and filters due to the change from lowercase naming to uppercase. This impacted our existing old datasets when I clicked "Sync Columns From Source."

Is there any guidance on how to make this backward compatible with existing assets? @betodealmeida @Vitor-Avila

**Comment by crysensible** [comment]
Here was the fix for anyone who encounters this: https://status.preset.io/pages/incident/5e83b80830610b04c39e898b/64b5823a267e5c053c6ed085

**Comment by rusackas** [comment]
We're planning to fix forward on this repo. Please stay tuned. If you have an issue with this PR, you can revert it locally or open a PR to revert it for everyone if there are stronger feelings.
